### PR TITLE
Search for cross references between sections

### PR DESCRIPTION
**Pull request by wbowling**
### Description/Motivation/Screenshots ###
Fixes https://github.com/hugsy/gef/issues/326

gef-scan

### How Has This Been Tested? ###

| Architecture | Yes/No                   | Comments               |
|--------------|:------------------------:|------------------------|
| x86-32       | :heavy_check_mark: |                        |
| x86-64       | :heavy_check_mark: |                        |
| ARM          | :heavy_multiplication_x: |                        |
| AARCH64      | :heavy_multiplication_x: |                        |
| MIPS         | :heavy_multiplication_x: |                        |
| POWERPC      | :heavy_multiplication_x: |                        |
| SPARC        | :heavy_multiplication_x: |                        |

### Checklist ###

- [x] My code follows the code style of this project.
- [x] My change includes a change to the documentation, if required.
- [x] I have read and agree to the **CONTRIBUTING** document.

---
This change is [Reviewable](https://reviewable.io/reviews/hugsy/gef/332)

### CODE REVIEW

File: docs/commands/scan.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+## Command scan ##
+
+`scan` search for addresses belonging to one mapping that are located in
+another.
+
+![scan-example](https://i.imgur.com/4ScRvVc.png)
+
+`scan` requires two arguments, the first is the memory section that will be
+search and the second is what will be searched for.
```
Comment (Grazfather) [inline docs/commands/scan.md]: will be search*ed*

File: gef.py [hunk 2]
```diff
@@ -4834,6 +4834,59 @@ def import_structures(self, structs):
...
+        for sect in get_process_maps():
+            if belong in sect.path:
+                belong_sections += [(sect.page_start, sect.page_end)]
+            if location in sect.path:
+                location_sections += [(sect.page_start, sect.page_end)]
+
+        step = current_arch.ptrsize
+        fmt = "{}{}".format(endian_str(), "I" if step==4 else "Q")
+
+        for (bstart, bend) in belong_sections:
```
Comment (Grazfather) [inline gef.py]: parens aren't needed here.

File: gef.py [hunk 3]
```diff
@@ -4834,6 +4834,59 @@ def import_structures(self, structs):
...
+            try:
+                mem = read_memory(bstart, bend - bstart)
+            except gdb.MemoryError:
+                continue
+
+            for i in range(0, len(mem), step):
+                target = struct.unpack(fmt, mem[i:i+step])[0]
+                for (lstart, lend) in location_sections:
```
Comment (Grazfather) [inline gef.py]: no parens

File: gef.py [hunk 4]
```diff
@@ -4834,6 +4834,59 @@ def import_structures(self, structs):
         return
 
 
+@register_command
+class ScanSectionCommand(GenericCommand):
+    """Search for addresses belonging to one mapping that are located in another."""
+
+    _cmdline_ = "scan"
+    _syntax_  = "{:s} BELONG LOCATION".format(_cmdline_)
```
Comment (Grazfather) [inline gef.py]: I think BELONG and LOCATION are weird names for the args.

Maybe something like section and target? Can you think of something else?

File: gef.py [hunk 5]
```diff
@@ -4834,6 +4834,59 @@ def import_structures(self, structs):
...
+
+        if location == "binary":
+            location = get_filepath()
+
+        belong_sections = []
+        location_sections = []
+
+        for sect in get_process_maps():
+            if belong in sect.path:
+                belong_sections += [(sect.page_start, sect.page_end)]
```
Comment (Grazfather) [inline gef.py]: let's just use `.append`. It's marginally faster (last I checked)

**Comment by Grazfather** [comment]
Please retarget the dev branch

**Comment by hugsy** [comment]
@wbowling Great job. For all new commands we ask detailed documentation, along with some unit tests. Would be great if you added that to your PR.

Cheers

**Comment by Grazfather** [comment]
I can help you write a test. I think it'd just be a matter of loading up a binary (we compile a few already for testing purposes, see _tests/binaries_, and then doing a search (libc in stack e.g.) and make sure we see a result.

File: gef.py (hunk 4 above)
Comment (wbowling) [reply gef.py]: Naming things is hard. I wasn’t too happy with them but the others sounded ambiguous. I’ll have another go

**Comment by Grazfather** [comment]
LGTM!

**Comment by Grazfather** [comment]
Tests fail for me
```
gef➤  scan libc stack
[+] Searching for addresses in 'libc' that point to 'stack'
libc-2.23.so: 0x00007ffff7dd23d0│+0x13d0: 0x00007fffffffe896  →  "heap.out"
libc-2.23.so: 0x00007ffff7dd23d8│+0x13d8: 0x00007fffffffe878  →  "/root/code/gef/tests/binaries/heap.out"
gef➤  scan libc binary
[+] Searching for addresses in 'libc' that point to 'binary'
```

**Comment by wbowling** [comment]
> So for one it looks like you're counting on the test binary not being compiled with PIE, and 2 you might not be using the args in the correct order.

Ah yeah I forgot the 18.04 defaults to PIE, was testing on 16.04 where it doesn't.

**Comment by wbowling** [comment]
Changed the test binaries to ones that have pie explicitly set/unset.

First test should find `program_invocation_short_name` in libc which points to the stack, the second should find `__libc_start_main` in the binary's got which points to libc

**Comment by hugsy** [comment]
Fantastic job @wbowling , great feature ! And to @Grazfather for the review as well.

**Pull request merged**